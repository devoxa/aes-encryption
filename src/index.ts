import * as crypto from 'crypto'

export function encrypt(key: string, input: string): string {
  const keyBuffer = Buffer.from(key)
  const iv = crypto.randomBytes(16)

  const cipher = crypto.createCipheriv('aes-256-gcm', keyBuffer, iv)

  const encrypted = Buffer.concat([cipher.update(input, 'utf8'), cipher.final()])
  const tag = cipher.getAuthTag()

  return Buffer.concat([iv, tag, encrypted]).toString('base64')
}

export function decrypt(key: string, input: string): string {
  const keyBuffer = Buffer.from(key)
  const inputBuffer = Buffer.from(input, 'base64')

  const iv = inputBuffer.subarray(0, 16)
  const tag = inputBuffer.subarray(16, 32)
  const text = inputBuffer.subarray(32)

  const decipher = crypto.createDecipheriv('aes-256-gcm', keyBuffer, iv)
  decipher.setAuthTag(tag)

  const decrypted = decipher.update(text).toString() + decipher.final('utf8')

  return decrypted
}
