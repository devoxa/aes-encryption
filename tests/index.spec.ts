import { encrypt, decrypt } from '../src/index'

const KEY = 'gdPfTP7h1hcs1ySGp0vcFwIZVpWfUSqJ'

describe('aes-encryption', () => {
  it('can encrypt and decrypt a string', () => {
    expect(decrypt(KEY, encrypt(KEY, 'FooBar'))).toEqual('FooBar')
    expect(decrypt(KEY, encrypt(KEY, '👋 🤚 🖐'))).toEqual('👋 🤚 🖐')
  })

  it('does not produce the same output twice', () => {
    const first = encrypt(KEY, 'FooBar')
    const second = encrypt(KEY, 'FooBar')

    expect(first === second).toEqual(false)
    expect(decrypt(KEY, first)).toEqual('FooBar')
    expect(decrypt(KEY, second)).toEqual('FooBar')
  })

  it('errors for the wrong encryption key when decrypting', () => {
    const encrypted = encrypt(KEY, 'FooBar')

    expect(() =>
      decrypt('AAPfTP7h1hcs1ySGp0vcFwIZVpWfUSAA', encrypted)
    ).toThrowErrorMatchingSnapshot()
  })

  it('errors when a wrong key format is provided', () => {
    const encrypted = encrypt(KEY, 'FooBar')

    expect(() =>
      encrypt('AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA', 'FooBar')
    ).toThrowErrorMatchingSnapshot()
    expect(() => encrypt('AAAAAAAAAAAAAAAAAA', 'FooBar')).toThrowErrorMatchingSnapshot()
    // @ts-expect-error key has to be a string
    expect(() => encrypt([1, 2, 3], 'FooBar')).toThrowErrorMatchingSnapshot()
    expect(() =>
      decrypt('AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA', encrypted)
    ).toThrowErrorMatchingSnapshot()
    expect(() => decrypt('AAAAAAAAAAAAAAAAAA', encrypted)).toThrowErrorMatchingSnapshot()
    // @ts-expect-error key has to be a string
    expect(() => encrypt([1, 2, 3], 'FooBar')).toThrowErrorMatchingSnapshot()
  })

  it('errors when a wrong input format is provided', () => {
    const encrypted = encrypt(KEY, 'FooBar')

    // @ts-expect-error input has to be a string
    expect(() => encrypt(KEY, null)).toThrowErrorMatchingSnapshot()
    // @ts-expect-error input has to be a string
    expect(() => encrypt(KEY, [1, 2, 3])).toThrowErrorMatchingSnapshot()
    expect(() => decrypt(KEY, 'foo' + encrypted)).toThrowErrorMatchingSnapshot()
    // @ts-expect-error input has to be a string
    expect(() => decrypt(KEY, [encrypted])).toThrowErrorMatchingSnapshot()
  })
})
