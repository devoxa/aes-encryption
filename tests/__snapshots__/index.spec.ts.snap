// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aes-encryption errors for the wrong encryption key when decrypting 1`] = `"Unsupported state or unable to authenticate data"`;

exports[`aes-encryption errors when a wrong input format is provided 1`] = `"The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received null"`;

exports[`aes-encryption errors when a wrong input format is provided 2`] = `"The "data" argument must be of type string or an instance of Buffer, TypedArray, or DataView. Received an instance of Array"`;

exports[`aes-encryption errors when a wrong input format is provided 3`] = `"Unsupported state or unable to authenticate data"`;

exports[`aes-encryption errors when a wrong input format is provided 4`] = `"Invalid authentication tag length: 0"`;

exports[`aes-encryption errors when a wrong key format is provided 1`] = `"Invalid key length"`;

exports[`aes-encryption errors when a wrong key format is provided 2`] = `"Invalid key length"`;

exports[`aes-encryption errors when a wrong key format is provided 3`] = `"Invalid key length"`;

exports[`aes-encryption errors when a wrong key format is provided 4`] = `"Invalid key length"`;

exports[`aes-encryption errors when a wrong key format is provided 5`] = `"Invalid key length"`;

exports[`aes-encryption errors when a wrong key format is provided 6`] = `"Invalid key length"`;
